PurchasesCoreSwift/Caching/DeviceCache.swift

```diff
@@ -0,0 +1,497 @@
+                    if !attribute.isSynced {
+                        attributesForUser[attributeKey] = attribute
+                    }
+                }
+                if attributesForUser.count > 0 {
+                    attributes[appUserID] = attributesForUser
+                }
+            }
+            return attributes
+        }
+    }
+
+    @objc open func deleteAttributesIfSynced(appUserID: String) {
+        writeCache {
+            guard self.threadUnsafeUnsyncedAttributesByKey(appUserID: appUserID).isEmpty else {
+                return
+            }
+            var groupedAttributes = self.threadUnsafeStoredAttributes
+            let attibutesForAppUserID = groupedAttributes.removeValue(forKey: appUserID)
+            guard attibutesForAppUserID != nil else {
+                Logger.warn("Attempt to delete synced attributes for \(appUserID), but there were none to delete")
+                return
+            }
+            self.userDefaults.setValue(groupedAttributes, forKey: CacheKeys.subscriberAttributes)
+        }
+    }
+
+    // MARK: - attribution
+
+    @objc public func latestNetworkAndAdvertisingIdsSent(appUserID: String) -> [String: Any] {
+        readCache {
+            let key = CacheKeyBases.attributionDataDefaults + appUserID
+            return self.userDefaults.object(forKey: key) as? [String: Any] ?? [:]
+        }
+    }
+
+    @objc(setLatestNetworkAndAdvertisingIdsSent:forAppUserID:)
+    public func set(latestNetworkAndAdvertisingIdsSent: [String: Any], appUserID: String) {
+        writeCache {
+            self.userDefaults.setValue(latestNetworkAndAdvertisingIdsSent,
+                                       forKey: CacheKeyBases.attributionDataDefaults + appUserID)
+        }
+    }
+
+    @objc open func clearLatestNetworkAndAdvertisingIdsSent(appUserID: String) {
+        writeCache {
+            self.userDefaults.removeObject(forKey: CacheKeyBases.attributionDataDefaults + appUserID)
+        }
+    }
+
+    private func cacheDurationInSeconds(isAppBackgrounded: Bool) -> Double {
+        return isAppBackgrounded ? cacheDurationInSecondsInBackground : cacheDurationInSecondsInForeground
+    }
+
+    // MARK: - Helper functions
+
+    // Uses the accessQueue to synchronously write to the UserDefaults.
+    // Warning: do NOT nest calls to `writeCache`, it will deadlock.
+    // Only make calls to functions starting with "threadUnsafe" as those are guaranteed to not have any locking
+    // mechanisms.
+    private func writeCache(block: @escaping () -> Void) {
+        // .barrier is not needed here because we're using `.sync` instead of the normal .async multi-reader
+        // single-writer dispatch queue synchronization pattern. However, because we're using this as a way
+        // to control access to the datasource, passing `.barrier` flag is meant to help signal that intent.
+        accessQueue.sync(flags: .barrier) {
```

Comment:
would a serial queue work?

Reply:
Updated with a serial queue 😄

Reply:
if we're using a serial queue, `.barrier` no longer does anything, right? if not, we should remove the param and the comment

Reply:
So, technically yeah, totally right and I agree. I'm thinking about future people looking at the code and changing it. The `.async(flags: .barrier)` pattern is pretty standard access control now, and I kinda think leaving this comment plus the `.barrier` flag to let people know "hey, this line of code is actually very important".  It's a signal to folks to reallllly consider what they are doing because it's pretty confusing.

So, I'm not sure what to do. If this were my own project, I'd leave this all here to remind my dumb future self, cause I'd look at the `.barrier` and go, wait, wut? and then think more.

Thoughts?